src/adventure/adventuretracker.cpp

```diff
@@ -0,0 +1,164 @@
+#include <QJsonDocument>
+#include <QJsonObject>
+
+#include "adventuretracker.h"
+#include "parser/parserutils.h"
+
+#include <QDebug>
+
+AdventureTracker::AdventureTracker(GameObserver &observer, QObject *const parent)
+    : QObject{parent}
+    , m_observer{observer}
+{
+    connect(&m_observer,
+            &GameObserver::sig_sentToUserText,
+            this,
+            &AdventureTracker::slot_onUserText);
+
+    connect(&m_observer,
+            &GameObserver::sig_sentToUserGmcp,
+            this,
+            &AdventureTracker::slot_onUserGmcp);
+}
+
+AdventureTracker::~AdventureTracker() {}
+
+void AdventureTracker::slot_onUserText(const QByteArray &ba)
+{
+    QString str = QString::fromLatin1(ba).trimmed();
+    ParserUtils::removeAnsiMarksInPlace(str);
+
+    // We keep track of the N most recent lines so that we can parse with the
+    // context of a sliding window. Array indices are interpeted backwards,
+    // so [0] is the last line received, [1] is the second-to-last, etc.
+    // Must create our own QString copy for this, since the `ba` passed above
+    // is reused. So shfit the array and add a copy.
+    for (size_t i = m_lastLines.size(); i > 0; i--) {
+        m_lastLines[i] = m_lastLines[i - 1];
```

Comment:
Aren't we leaking memory here? We're never deleting the QString that was created N + 1 copies ago.

Reply:
Yes we are leaking. Would you rather I use a list with push/pop or just explicitly delete the N+1 string here?

Reply:
I've replaced this entirely with stateful parser and no longer maintain a sliding window as a separate copy.